// @cardano-sdk/ogmios/dist/cjs/ogmiosToCore/block.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.block = exports.blockHeader = void 0;
const util_1 = require("@cardano-sdk/util");
const client_1 = require("@cardano-ogmios/client");
const Crypto = __importStar(require("@cardano-sdk/crypto"));
const core_1 = require("@cardano-sdk/core");
const tx_1 = require("./tx");
const getBlockAndKind = (block) => {
    let propName = 'alonzo';
    if ((0, client_1.isAllegraBlock)(block))
        propName = 'allegra';
    if ((0, client_1.isAlonzoBlock)(block))
        propName = 'alonzo';
    if ((0, client_1.isBabbageBlock)(block))
        propName = 'babbage';
    if ((0, client_1.isByronBlock)(block))
        propName = 'byron';
    if ((0, client_1.isMaryBlock)(block))
        propName = 'mary';
    if ((0, client_1.isShelleyBlock)(block))
        propName = 'shelley';
    switch (propName) {
        case 'allegra':
            return { block: block.allegra, kind: 'allegra' };
        case 'alonzo':
            return { block: block.alonzo, kind: 'alonzo' };
        case 'babbage':
            return { block: block.babbage, kind: 'babbage' };
        case 'byron':
            return (0, client_1.isByronStandardBlock)(block) ? { block: block.byron, kind: 'byron' } : null;
        case 'mary':
            return { block: block.mary, kind: 'mary' };
        case 'shelley':
            return { block: block.shelley, kind: 'shelley' };
        default: {
            const _exhaustiveCheck = propName;
            return _exhaustiveCheck;
        }
    }
};
const mapBlockHeight = (block) => block.header.blockHeight;
const mapBlockSlot = (block) => block.header.slot;
const mapPreviousBlock = (block) => block.header.prevHash !== 'genesis' ? core_1.Cardano.BlockId(block.header.prevHash) : undefined;
const mapByronHash = (block) => core_1.Cardano.BlockId(block.hash);
const mapByronTotalOutputs = (block) => util_1.BigIntMath.sum(block.body.txPayload.map(({ body: { outputs } }) => util_1.BigIntMath.sum(outputs.map(({ value: { coins } }) => coins))));
const mapByronTxCount = (block) => block.body.txPayload.length;
const mapCommonTxCount = (block) => block.body.length;
const mapCommonHash = (block) => core_1.Cardano.BlockId(block.headerHash);
const mapCommonTotalOutputs = (block) => util_1.BigIntMath.sum(block.body.map(({ body: { outputs } }) => util_1.BigIntMath.sum(outputs.map(({ value: { coins } }) => coins))));
const mapCommonBlockSize = (block) => core_1.Cardano.BlockSize(block.header.blockSize);
const mapCommonFees = (block) => block.body.map(({ body: { fee } }) => fee).reduce((prev, current) => prev + current, 0n);
const mapCommonVrf = (block) => core_1.Cardano.VrfVkBech32FromBase64(block.header.issuerVrf);
const mapCommonSlotLeader = (block) => Crypto.Ed25519PublicKeyHex(block.header.issuerVk);
const mapStandardBlockHeader = (block) => ({
    blockNo: core_1.Cardano.BlockNo(mapBlockHeight(block)),
    hash: mapByronHash(block),
    slot: core_1.Cardano.Slot(mapBlockSlot(block))
});
const mapCommonBlockHeader = (block) => ({
    blockNo: core_1.Cardano.BlockNo(mapBlockHeight(block)),
    hash: mapCommonHash(block),
    slot: core_1.Cardano.Slot(mapBlockSlot(block))
});
const mapByronBlock = (block) => ({
    body: (0, tx_1.mapByronBlockBody)(block),
    fees: undefined,
    header: mapStandardBlockHeader(block),
    issuerVk: undefined,
    previousBlock: mapPreviousBlock(block),
    size: undefined,
    totalOutput: mapByronTotalOutputs(block),
    txCount: mapByronTxCount(block),
    vrf: undefined
});
const mapCommonBlock = (block, kind) => ({
    body: (0, tx_1.mapCommonBlockBody)(block, kind),
    fees: mapCommonFees(block),
    header: mapCommonBlockHeader(block),
    issuerVk: mapCommonSlotLeader(block),
    previousBlock: mapPreviousBlock(block),
    size: mapCommonBlockSize(block),
    totalOutput: mapCommonTotalOutputs(block),
    txCount: mapCommonTxCount(block),
    vrf: mapCommonVrf(block)
});
const mapBlock = (ogmiosBlock, mapStandardBlock, mapOtherBlock) => {
    const b = getBlockAndKind(ogmiosBlock);
    if (!b)
        return null;
    switch (b.kind) {
        case 'byron': {
            return mapStandardBlock(b.block);
        }
        case 'babbage':
        case 'allegra':
        case 'alonzo':
        case 'mary':
        case 'shelley': {
            return mapOtherBlock(b.block, b.kind);
        }
        default: {
            const _exhaustiveCheck = b;
            return _exhaustiveCheck;
        }
    }
};
const blockHeader = (ogmiosBlock) => mapBlock(ogmiosBlock, mapStandardBlockHeader, mapCommonBlockHeader);
exports.blockHeader = blockHeader;
const block = (ogmiosBlock) => mapBlock(ogmiosBlock, mapByronBlock, mapCommonBlock);
exports.block = block;
