// @cardano-sdk/ogmios/dist/cjs/ogmiosToCore/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isShelleyTx = exports.isMaryOrAbove = exports.isAlonzoOrAbove = exports.isRequireNOf = exports.isStartsAt = exports.isExpiresAt = exports.isRequireAnyOf = exports.isRequireAllOf = exports.isPlutusV2Script = exports.isPlutusV1Script = exports.isNativeScript = exports.BYRON_TX_FEE_CONSTANT = exports.BYRON_TX_FEE_COEFFICIENT = void 0;
exports.BYRON_TX_FEE_COEFFICIENT = 43946000000;
exports.BYRON_TX_FEE_CONSTANT = 155381000000000;
const isNativeScript = (script) => 'native' in script;
exports.isNativeScript = isNativeScript;
const isPlutusV1Script = (script) => 'plutus:v1' in script;
exports.isPlutusV1Script = isPlutusV1Script;
const isPlutusV2Script = (script) => 'plutus:v2' in script;
exports.isPlutusV2Script = isPlutusV2Script;
const isRequireAllOf = (nativeScript) => typeof nativeScript === 'object' && 'all' in nativeScript;
exports.isRequireAllOf = isRequireAllOf;
const isRequireAnyOf = (nativeScript) => typeof nativeScript === 'object' && 'any' in nativeScript;
exports.isRequireAnyOf = isRequireAnyOf;
const isExpiresAt = (nativeScript) => typeof nativeScript === 'object' && 'expiresAt' in nativeScript;
exports.isExpiresAt = isExpiresAt;
const isStartsAt = (nativeScript) => typeof nativeScript === 'object' && 'startsAt' in nativeScript;
exports.isStartsAt = isStartsAt;
const isRequireNOf = (nativeScript) => typeof nativeScript === 'object' && !Number.isNaN(Number(Object.keys(nativeScript)[0]));
exports.isRequireNOf = isRequireNOf;
const isAlonzoOrAbove = (kind) => kind === 'babbage' || kind === 'alonzo';
exports.isAlonzoOrAbove = isAlonzoOrAbove;
const isMaryOrAbove = (kind) => (0, exports.isAlonzoOrAbove)(kind) || kind === 'mary';
exports.isMaryOrAbove = isMaryOrAbove;
const isShelleyTx = (kind) => kind === 'shelley';
exports.isShelleyTx = isShelleyTx;
