// @cardano-sdk/core/dist/esm/Cardano/types/Block.js
import * as Crypto from '@cardano-sdk/crypto';
import { CML } from '../../CML/CML';
import { InvalidStringError, typedBech32 } from '@cardano-sdk/util';
import { PoolId } from './StakePool/primitives';
export const BlockSize = (value) => value;
export const BlockNo = (value) => value;
export const EpochNo = (value) => value;
export const Slot = (value) => value;
export const BlockId = (value) => Crypto.Hash32ByteBase16(value);
export const VrfVkBech32 = (value) => typedBech32(value, 'vrf_vk', 52);
export const GenesisDelegate = (value) => {
    if (/ShelleyGenesis-[\da-f]{16}/.test(value)) {
        return value;
    }
    return Crypto.Hash28ByteBase16(value);
};
export const SlotLeader = (value) => {
    try {
        return PoolId(value);
    }
    catch {
        try {
            return GenesisDelegate(value);
        }
        catch (error) {
            throw new InvalidStringError('Expected either PoolId or GenesisDelegate', error);
        }
    }
};
export const VrfVkBech32FromBase64 = (value) => VrfVkBech32(CML.VRFVKey.from_bytes(Buffer.from(value, 'base64')).to_bech32('vrf_vk'));
