// @cardano-sdk/core/dist/cjs/CardanoNode/types/CardanoNodeErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxSubmissionErrors = exports.UnknownTxSubmissionError = exports.CardanoClientErrors = exports.UnknownCardanoNodeError = exports.NotInitializedError = void 0;
const client_1 = require("@cardano-ogmios/client");
const util_1 = require("@cardano-sdk/util");
const ts_custom_error_1 = require("ts-custom-error");
class NotInitializedError extends ts_custom_error_1.CustomError {
    constructor(methodName, moduleName) {
        super(`${methodName} cannot be called until ${moduleName} is initialized`);
    }
}
exports.NotInitializedError = NotInitializedError;
class UnknownCardanoNodeError extends util_1.ComposableError {
    constructor(innerError) {
        super('Unknown CardanoNode error', innerError);
    }
}
exports.UnknownCardanoNodeError = UnknownCardanoNodeError;
exports.CardanoClientErrors = {
    AcquirePointNotOnChainError: client_1.AcquirePointNotOnChainError,
    AcquirePointTooOldError: client_1.AcquirePointTooOldError,
    ConnectionError: client_1.WebSocketClosed,
    EraMismatchError: client_1.EraMismatchError,
    IntersectionNotFoundError: client_1.IntersectionNotFoundError,
    QueryUnavailableInCurrentEraError: client_1.QueryUnavailableInCurrentEraError,
    ServerNotReady: client_1.ServerNotReady,
    TipIsOriginError: client_1.TipIsOriginError,
    UnknownResultError: client_1.UnknownResultError
};
class UnknownTxSubmissionError extends util_1.ComposableError {
    constructor(innerError) {
        super('Unknown submission error', innerError);
    }
}
exports.UnknownTxSubmissionError = UnknownTxSubmissionError;
const ogmiosSubmissionErrors = client_1.TxSubmission.submissionErrors.errors;
exports.TxSubmissionErrors = {
    AddressAttributesTooLargeError: ogmiosSubmissionErrors.AddressAttributesTooLarge.Error,
    AlreadyDelegatingError: ogmiosSubmissionErrors.AlreadyDelegating.Error,
    BadInputsError: ogmiosSubmissionErrors.BadInputs.Error,
    CollateralHasNonAdaAssetsError: ogmiosSubmissionErrors.CollateralHasNonAdaAssets.Error,
    CollateralIsScriptError: ogmiosSubmissionErrors.CollateralIsScript.Error,
    CollateralTooSmallError: ogmiosSubmissionErrors.CollateralTooSmall.Error,
    CollectErrorsError: ogmiosSubmissionErrors.CollectErrors.Error,
    DelegateNotRegisteredError: ogmiosSubmissionErrors.DelegateNotRegistered.Error,
    DuplicateGenesisVrfError: ogmiosSubmissionErrors.DuplicateGenesisVrf.Error,
    EraMismatchError: ogmiosSubmissionErrors.EraMismatch.Error,
    ExecutionUnitsTooLargeError: ogmiosSubmissionErrors.ExecutionUnitsTooLarge.Error,
    ExpiredUtxoError: ogmiosSubmissionErrors.ExpiredUtxo.Error,
    ExtraDataMismatchError: ogmiosSubmissionErrors.ExtraDataMismatch.Error,
    ExtraRedeemersError: ogmiosSubmissionErrors.ExtraRedeemers.Error,
    ExtraScriptWitnessesError: ogmiosSubmissionErrors.ExtraScriptWitnesses.Error,
    FeeTooSmallError: ogmiosSubmissionErrors.FeeTooSmall.Error,
    InsufficientFundsForMirError: ogmiosSubmissionErrors.InsufficientFundsForMir.Error,
    InsufficientGenesisSignaturesError: ogmiosSubmissionErrors.InsufficientGenesisSignatures.Error,
    InvalidMetadataError: ogmiosSubmissionErrors.InvalidMetadata.Error,
    InvalidWitnessesError: ogmiosSubmissionErrors.InvalidWitnesses.Error,
    MalformedReferenceScriptsError: ogmiosSubmissionErrors.MalformedReferenceScripts.Error,
    MalformedScriptWitnessesError: ogmiosSubmissionErrors.MalformedScriptWitnesses.Error,
    MirNegativeTransferError: ogmiosSubmissionErrors.MirNegativeTransfer.Error,
    MirNegativeTransferNotCurrentlyAllowedError: ogmiosSubmissionErrors.MirNegativeTransferNotCurrentlyAllowed.Error,
    MirProducesNegativeUpdateError: ogmiosSubmissionErrors.MirProducesNegativeUpdate.Error,
    MirTransferNotCurrentlyAllowedError: ogmiosSubmissionErrors.MirTransferNotCurrentlyAllowed.Error,
    MissingAtLeastOneInputUtxoError: ogmiosSubmissionErrors.MissingAtLeastOneInputUtxo.Error,
    MissingCollateralInputsError: ogmiosSubmissionErrors.MissingCollateralInputs.Error,
    MissingDatumHashesForInputsError: ogmiosSubmissionErrors.MissingDatumHashesForInputs.Error,
    MissingRequiredDatumsError: ogmiosSubmissionErrors.MissingRequiredDatums.Error,
    MissingRequiredRedeemersError: ogmiosSubmissionErrors.MissingRequiredRedeemers.Error,
    MissingRequiredSignaturesError: ogmiosSubmissionErrors.MissingRequiredSignatures.Error,
    MissingScriptWitnessesError: ogmiosSubmissionErrors.MissingScriptWitnesses.Error,
    MissingTxMetadataError: ogmiosSubmissionErrors.MissingTxMetadata.Error,
    MissingTxMetadataHashError: ogmiosSubmissionErrors.MissingTxMetadataHash.Error,
    MissingVkWitnessesError: ogmiosSubmissionErrors.MissingVkWitnesses.Error,
    NetworkMismatchError: ogmiosSubmissionErrors.NetworkMismatch.Error,
    NonGenesisVotersError: ogmiosSubmissionErrors.NonGenesisVoters.Error,
    OutputTooSmallError: ogmiosSubmissionErrors.OutputTooSmall.Error,
    OutsideForecastError: ogmiosSubmissionErrors.OutsideForecast.Error,
    OutsideOfValidityIntervalError: ogmiosSubmissionErrors.OutsideOfValidityInterval.Error,
    PoolCostTooSmallError: ogmiosSubmissionErrors.PoolCostTooSmall.Error,
    PoolMetadataHashTooBigError: ogmiosSubmissionErrors.PoolMetadataHashTooBig.Error,
    ProtocolVersionCannotFollowError: ogmiosSubmissionErrors.ProtocolVersionCannotFollow.Error,
    RewardAccountNotEmptyError: ogmiosSubmissionErrors.RewardAccountNotEmpty.Error,
    RewardAccountNotExistingError: ogmiosSubmissionErrors.RewardAccountNotExisting.Error,
    ScriptWitnessNotValidatingError: ogmiosSubmissionErrors.ScriptWitnessNotValidating.Error,
    StakeKeyAlreadyRegisteredError: ogmiosSubmissionErrors.StakeKeyAlreadyRegistered.Error,
    StakeKeyNotRegisteredError: ogmiosSubmissionErrors.StakeKeyNotRegistered.Error,
    StakePoolNotRegisteredError: ogmiosSubmissionErrors.StakePoolNotRegistered.Error,
    TooLateForMirError: ogmiosSubmissionErrors.TooLateForMir.Error,
    TooManyAssetsInOutputError: ogmiosSubmissionErrors.TooManyAssetsInOutput.Error,
    TooManyCollateralInputsError: ogmiosSubmissionErrors.TooManyCollateralInputs.Error,
    TotalCollateralMismatchError: ogmiosSubmissionErrors.TotalCollateralMismatch.Error,
    TriesToForgeAdaError: ogmiosSubmissionErrors.TriesToForgeAda.Error,
    TxMetadataHashMismatchError: ogmiosSubmissionErrors.TxMetadataHashMismatch.Error,
    TxTooLargeError: ogmiosSubmissionErrors.TxTooLarge.Error,
    UnknownGenesisKeyError: ogmiosSubmissionErrors.UnknownGenesisKey.Error,
    UnknownOrIncompleteWithdrawalsError: ogmiosSubmissionErrors.UnknownOrIncompleteWithdrawals.Error,
    UnspendableDatumsError: ogmiosSubmissionErrors.UnspendableDatums.Error,
    UnspendableScriptInputsError: ogmiosSubmissionErrors.UnspendableScriptInputs.Error,
    UpdateWrongEpochError: ogmiosSubmissionErrors.UpdateWrongEpoch.Error,
    ValidationTagMismatchError: ogmiosSubmissionErrors.ValidationTagMismatch.Error,
    ValueNotConservedError: ogmiosSubmissionErrors.ValueNotConserved.Error,
    WrongCertificateTypeError: ogmiosSubmissionErrors.WrongCertificateType.Error,
    WrongPoolCertificateError: ogmiosSubmissionErrors.WrongPoolCertificate.Error,
    WrongRetirementEpochError: ogmiosSubmissionErrors.WrongRetirementEpoch.Error
};
