// @cardano-sdk/ogmios/dist/esm/ogmiosToCore/util.js
export const BYRON_TX_FEE_COEFFICIENT = 43946000000;
export const BYRON_TX_FEE_CONSTANT = 155381000000000;
export const isNativeScript = (script) => 'native' in script;
export const isPlutusV1Script = (script) => 'plutus:v1' in script;
export const isPlutusV2Script = (script) => 'plutus:v2' in script;
export const isRequireAllOf = (nativeScript) => typeof nativeScript === 'object' && 'all' in nativeScript;
export const isRequireAnyOf = (nativeScript) => typeof nativeScript === 'object' && 'any' in nativeScript;
export const isExpiresAt = (nativeScript) => typeof nativeScript === 'object' && 'expiresAt' in nativeScript;
export const isStartsAt = (nativeScript) => typeof nativeScript === 'object' && 'startsAt' in nativeScript;
export const isRequireNOf = (nativeScript) => typeof nativeScript === 'object' && !Number.isNaN(Number(Object.keys(nativeScript)[0]));
export const isAlonzoOrAbove = (kind) => kind === 'babbage' || kind === 'alonzo';
export const isMaryOrAbove = (kind) => isAlonzoOrAbove(kind) || kind === 'mary';
export const isShelleyTx = (kind) => kind === 'shelley';
