// @cardano-sdk/core/dist/cjs/Cardano/types/Certificate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MirCertificatePot = exports.CertificateType = void 0;
var CertificateType;
(function (CertificateType) {
    CertificateType["StakeKeyRegistration"] = "StakeKeyRegistrationCertificate";
    CertificateType["StakeKeyDeregistration"] = "StakeKeyDeregistrationCertificate";
    CertificateType["PoolRegistration"] = "PoolRegistrationCertificate";
    CertificateType["PoolRetirement"] = "PoolRetirementCertificate";
    CertificateType["StakeDelegation"] = "StakeDelegationCertificate";
    CertificateType["MIR"] = "MirCertificate";
    CertificateType["GenesisKeyDelegation"] = "GenesisKeyDelegationCertificate";
})(CertificateType = exports.CertificateType || (exports.CertificateType = {}));
var MirCertificatePot;
(function (MirCertificatePot) {
    MirCertificatePot["Reserves"] = "reserve";
    MirCertificatePot["Treasury"] = "treasury";
})(MirCertificatePot = exports.MirCertificatePot || (exports.MirCertificatePot = {}));
