// @cardano-ogmios/client/dist/StateQuery/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownResultError = exports.TipIsOriginError = exports.QueryUnavailableInCurrentEraError = exports.EraMismatchError = exports.AcquirePointTooOldError = exports.AcquirePointNotOnChainError = void 0;
__exportStar(require("./queries"), exports);
__exportStar(require("./Query"), exports);
__exportStar(require("./StateQueryClient"), exports);
var errors_1 = require("../errors");
Object.defineProperty(exports, "AcquirePointNotOnChainError", { enumerable: true, get: function () { return errors_1.AcquirePointNotOnChainError; } });
Object.defineProperty(exports, "AcquirePointTooOldError", { enumerable: true, get: function () { return errors_1.AcquirePointTooOldError; } });
Object.defineProperty(exports, "EraMismatchError", { enumerable: true, get: function () { return errors_1.EraMismatchError; } });
Object.defineProperty(exports, "QueryUnavailableInCurrentEraError", { enumerable: true, get: function () { return errors_1.QueryUnavailableInCurrentEraError; } });
Object.defineProperty(exports, "TipIsOriginError", { enumerable: true, get: function () { return errors_1.TipIsOriginError; } });
Object.defineProperty(exports, "UnknownResultError", { enumerable: true, get: function () { return errors_1.UnknownResultError; } });
