// @polkadot/typegen/generate/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "generateDefaultConsts", {
  enumerable: true,
  get: function () {
    return _consts.generateDefaultConsts;
  }
});
Object.defineProperty(exports, "generateDefaultErrors", {
  enumerable: true,
  get: function () {
    return _errors.generateDefaultErrors;
  }
});
Object.defineProperty(exports, "generateDefaultEvents", {
  enumerable: true,
  get: function () {
    return _events.generateDefaultEvents;
  }
});
Object.defineProperty(exports, "generateDefaultInterface", {
  enumerable: true,
  get: function () {
    return _interfaceRegistry.generateDefaultInterface;
  }
});
Object.defineProperty(exports, "generateDefaultLookup", {
  enumerable: true,
  get: function () {
    return _lookup.generateDefaultLookup;
  }
});
Object.defineProperty(exports, "generateDefaultQuery", {
  enumerable: true,
  get: function () {
    return _query.generateDefaultQuery;
  }
});
Object.defineProperty(exports, "generateDefaultRpc", {
  enumerable: true,
  get: function () {
    return _rpc.generateDefaultRpc;
  }
});
Object.defineProperty(exports, "generateDefaultTsDef", {
  enumerable: true,
  get: function () {
    return _tsDef.generateDefaultTsDef;
  }
});
Object.defineProperty(exports, "generateDefaultTx", {
  enumerable: true,
  get: function () {
    return _tx.generateDefaultTx;
  }
});

var _consts = require("./consts.cjs");

var _errors = require("./errors.cjs");

var _events = require("./events.cjs");

var _interfaceRegistry = require("./interfaceRegistry.cjs");

var _lookup = require("./lookup.cjs");

var _query = require("./query.cjs");

var _rpc = require("./rpc.cjs");

var _tsDef = require("./tsDef.cjs");

var _tx = require("./tx.cjs");
