// @polkadot/typegen/util/formatting.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HEADER = void 0;
exports.exportInterface = exportInterface;
exports.exportType = exportType;
exports.formatType = formatType;

var _handlebars = _interopRequireDefault(require("handlebars"));

var _create = require("@polkadot/types/create");

var _types = require("@polkadot/types/create/types");

var _util = require("@polkadot/util");

var _file = require("./file.cjs");

var _imports = require("./imports.cjs");

// Copyright 2017-2021 @polkadot/typegen authors & contributors
// SPDX-License-Identifier: Apache-2.0

/* eslint-disable @typescript-eslint/no-unused-vars */
const NO_CODEC = ['Tuple', 'VecFixed'];

const HEADER = type => `// Auto-generated via \`yarn polkadot-types-from-${type}\`, do not edit\n/* eslint-disable */\n\n`;

exports.HEADER = HEADER;

_handlebars.default.registerPartial({
  footer: _handlebars.default.compile((0, _file.readTemplate)('footer')),
  header: _handlebars.default.compile((0, _file.readTemplate)('header'))
});

_handlebars.default.registerHelper({
  imports() {
    const {
      imports,
      types
    } = this;
    const defs = [{
      file: '@polkadot/types',
      types: [...Object.keys(imports.codecTypes).filter(name => !NO_CODEC.includes(name)), ...Object.keys(imports.extrinsicTypes), ...Object.keys(imports.genericTypes), ...Object.keys(imports.metadataTypes), ...Object.keys(imports.primitiveTypes)]
    }, {
      file: '@polkadot/types/types',
      types: Object.keys(imports.typesTypes)
    }, ...types];
    return [...defs].sort((a, b) => a.file.localeCompare(b.file)).reduce((result, {
      file,
      types
    }) => {
      return types.length ? `${result}import type { ${types.sort().join(', ')} } from '${file}';\n` : result;
    }, '');
  },

  trim(options) {
    return options.fn(this).trim();
  },

  upper(options) {
    return options.fn(this).toUpperCase();
  }

}); // helper to generate a `export interface <Name> extends <Base> {<Body>}

/** @internal */


function exportInterface(lookupIndex = -1, name = '', base, body = '') {
  // * @description extends [[${base}]]
  const doc = `/** @name ${name}${lookupIndex !== -1 ? ` (${lookupIndex})` : ''} */\n`;
  return `${doc}export interface ${name} extends ${base} {${body.length ? '\n' : ''}${body}}`;
} // helper to create an `export type <Name> = <Base>`
// but since we don't want type alias (TS doesn't preserve names) we use
// interface here.

/** @internal */


function exportType(lookupIndex = -1, name = '', base) {
  return exportInterface(lookupIndex, name, base);
}

function singleParamNotation(registry, wrapper, typeDef, definitions, imports, withShortcut) {
  const sub = typeDef.sub;
  (0, _imports.setImports)(definitions, imports, [wrapper, sub.lookupName]);
  return (0, _create.paramsNotation)(wrapper, sub.lookupName || formatType(registry, definitions, sub.type, imports, withShortcut));
}

function dualParamsNotation(registry, wrapper, typeDef, definitions, imports, withShortcut) {
  const [a, b] = typeDef.sub;
  (0, _imports.setImports)(definitions, imports, [wrapper, a.lookupName, b.lookupName]);
  return (0, _create.paramsNotation)(wrapper, [a.lookupName || formatType(registry, definitions, a.type, imports, withShortcut), b.lookupName || formatType(registry, definitions, b.type, imports, withShortcut)]);
}

const formatters = {
  [_types.TypeDefInfo.Compact]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'Compact', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.DoNotConstruct]: (registry, {
    lookupName
  }, definitions, imports, withShortcut) => {
    (0, _imports.setImports)(definitions, imports, ['DoNotConstruct']);
    return 'DoNotConstruct';
  },
  [_types.TypeDefInfo.Enum]: (registry, typeDef, definitions, imports, withShortcut) => {
    if (typeDef.lookupName) {
      return typeDef.lookupName;
    }

    throw new Error(`TypeDefInfo.Enum: Not implemented on ${(0, _util.stringify)(typeDef)}`);
  },
  [_types.TypeDefInfo.Int]: (registry, typeDef, definitions, imports, withShortcut) => {
    throw new Error(`TypeDefInfo.Int: Not implemented on ${(0, _util.stringify)(typeDef)}`);
  },
  [_types.TypeDefInfo.UInt]: (registry, typeDef, definitions, imports, withShortcut) => {
    throw new Error(`TypeDefInfo.UInt: Not implemented on ${(0, _util.stringify)(typeDef)}`);
  },
  [_types.TypeDefInfo.Null]: (registry, typeDef, definitions, imports, withShortcut) => {
    (0, _imports.setImports)(definitions, imports, ['Null']);
    return 'Null';
  },
  [_types.TypeDefInfo.Option]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'Option', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.Plain]: (registry, typeDef, definitions, imports, withShortcut) => {
    (0, _imports.setImports)(definitions, imports, [typeDef.type]);
    return typeDef.type;
  },
  [_types.TypeDefInfo.Range]: (registry, typeDef, definitions, imports, withShortcut) => {
    throw new Error(`TypeDefInfo.Range: Not implemented on ${(0, _util.stringify)(typeDef)}`);
  },
  [_types.TypeDefInfo.Set]: (registry, typeDef, definitions, imports, withShortcut) => {
    throw new Error(`TypeDefInfo.Set: Not implemented on ${(0, _util.stringify)(typeDef)}`);
  },
  [_types.TypeDefInfo.Si]: (registry, typeDef, definitions, imports, withShortcut) => {
    return formatType(registry, definitions, registry.lookup.getTypeDef(typeDef.type), imports, withShortcut);
  },
  [_types.TypeDefInfo.Struct]: (registry, typeDef, definitions, imports, withShortcut) => {
    if (typeDef.lookupName) {
      return typeDef.lookupName;
    }

    const sub = typeDef.sub;
    (0, _imports.setImports)(definitions, imports, ['Struct', ...sub.map(({
      lookupName
    }) => lookupName)]);
    return `{${withShortcut ? ' ' : '\n'}${sub.map(({
      lookupName,
      name,
      type
    }, index) => [name || `unknown${index}`, lookupName || formatType(registry, definitions, type, imports, withShortcut)]).map(([k, t]) => `${withShortcut ? '' : '    readonly '}${k}: ${t};`).join(withShortcut ? ' ' : '\n')}${withShortcut ? ' ' : '\n  '}} & Struct`;
  },
  [_types.TypeDefInfo.Tuple]: (registry, typeDef, definitions, imports, withShortcut) => {
    const sub = typeDef.sub;
    (0, _imports.setImports)(definitions, imports, ['ITuple', ...sub.map(({
      lookupName
    }) => lookupName)]); // `(a,b)` gets transformed into `ITuple<[a, b]>`

    return (0, _create.paramsNotation)('ITuple', `[${sub.map(({
      lookupName,
      type
    }) => lookupName || formatType(registry, definitions, type, imports, withShortcut)).join(', ')}]`);
  },
  [_types.TypeDefInfo.Vec]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'Vec', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.VecFixed]: (registry, typeDef, definitions, imports, withShortcut) => {
    const sub = typeDef.sub;

    if (sub.type === 'u8') {
      (0, _imports.setImports)(definitions, imports, ['U8aFixed']);
      return 'U8aFixed';
    }

    return singleParamNotation(registry, 'Vec', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.BTreeMap]: (registry, typeDef, definitions, imports, withShortcut) => {
    return dualParamsNotation(registry, 'BTreeMap', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.BTreeSet]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'BTreeSet', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.HashMap]: (registry, typeDef, definitions, imports, withShortcut) => {
    return dualParamsNotation(registry, 'HashMap', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.Linkage]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'Linkage', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.Result]: (registry, typeDef, definitions, imports, withShortcut) => {
    return dualParamsNotation(registry, 'Result', typeDef, definitions, imports, withShortcut);
  },
  [_types.TypeDefInfo.WrapperOpaque]: (registry, typeDef, definitions, imports, withShortcut) => {
    return singleParamNotation(registry, 'WrapperOpaque', typeDef, definitions, imports, withShortcut);
  }
};
/**
 * Correctly format a given type
 */

/** @internal */
// eslint-disable-next-line @typescript-eslint/ban-types

function formatType(registry, definitions, type, imports, withShortcut = false) {
  let typeDef;

  if ((0, _util.isString)(type)) {
    const _type = type.toString(); // If type is "unorthodox" (i.e. `{ something: any }` for an Enum input or `[a | b | c, d | e | f]` for a Tuple's similar types),
    // we return it as-is


    if (withShortcut && /(^{.+:.+})|^\([^,]+\)|^\(.+\)\[\]|^\[.+\]/.exec(_type) && !/\[\w+;\w+\]/.exec(_type)) {
      return _type;
    }

    typeDef = (0, _create.getTypeDef)(type);
  } else {
    typeDef = type;
  }

  (0, _imports.setImports)(definitions, imports, [typeDef.lookupName || typeDef.type]);
  return formatters[typeDef.info](registry, typeDef, definitions, imports, withShortcut);
}
