// @cardano-sdk/core/dist/cjs/util/slotCalc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSlotEpochInfoCalc = exports.createSlotTimeCalc = exports.createSlotEpochCalc = exports.EraSummaryError = void 0;
const ts_custom_error_1 = require("ts-custom-error");
const Cardano_1 = require("../Cardano");
const groupBy_1 = __importDefault(require("lodash/groupBy"));
const last_1 = __importDefault(require("lodash/last"));
const memoize_1 = __importDefault(require("lodash/memoize"));
const orderBy_1 = __importDefault(require("lodash/orderBy"));
class EraSummaryError extends ts_custom_error_1.CustomError {
}
exports.EraSummaryError = EraSummaryError;
const createSlotEpochCalcImpl = (eraSummaries) => {
    const eraSummariesWithoutSkippedEras = Object.values((0, groupBy_1.default)(eraSummaries, 'start.slot')).map(last_1.default);
    const eraSummariesAsc = (0, orderBy_1.default)(eraSummariesWithoutSkippedEras, ({ start }) => start.slot);
    return (slotNo) => {
        const relevantEraSummariesAsc = (0, orderBy_1.default)(eraSummariesAsc.filter(({ start }) => start.slot <= slotNo), ({ start }) => start.slot);
        if (relevantEraSummariesAsc.length === 0) {
            throw new EraSummaryError(`No EraSummary for slot ${slotNo} found`);
        }
        let epochNo = 0;
        let currentEraSummary;
        for (let i = 0; i < relevantEraSummariesAsc.length; i++) {
            currentEraSummary = relevantEraSummariesAsc[i];
            const nextEraSummary = relevantEraSummariesAsc[i + 1];
            epochNo += Math.floor(((nextEraSummary?.start.slot || slotNo) - currentEraSummary.start.slot) /
                currentEraSummary.parameters.epochLength);
        }
        return { epochEraSummary: currentEraSummary, epochNo };
    };
};
exports.createSlotEpochCalc = (0, memoize_1.default)((eraSummaries) => {
    const calc = createSlotEpochCalcImpl(eraSummaries);
    return (slotNo) => (0, Cardano_1.EpochNo)(calc(slotNo).epochNo);
});
const createSlotTimeCalc = (eraSummaries) => {
    const eraSummariesDesc = (0, orderBy_1.default)(eraSummaries, ({ start }) => start.slot, 'desc');
    return (slotNo) => {
        const activeEraSummary = eraSummariesDesc.find(({ start }) => start.slot <= slotNo);
        if (!activeEraSummary) {
            throw new EraSummaryError(`No EraSummary for slot ${slotNo} found`);
        }
        return new Date(activeEraSummary.start.time.getTime() +
            (slotNo - activeEraSummary.start.slot) * activeEraSummary.parameters.slotLength);
    };
};
exports.createSlotTimeCalc = createSlotTimeCalc;
const createSlotEpochInfoCalc = (eraSummaries) => {
    const slotTimeCalc = (0, exports.createSlotTimeCalc)(eraSummaries);
    const epochCalc = createSlotEpochCalcImpl(eraSummaries);
    return (slot) => {
        const { epochNo, epochEraSummary } = epochCalc(slot);
        const firstSlot = epochEraSummary.start.slot +
            Math.floor((slot - epochEraSummary.start.slot) / epochEraSummary.parameters.epochLength) *
                epochEraSummary.parameters.epochLength;
        const lastSlot = firstSlot + epochEraSummary.parameters.epochLength - 1;
        return {
            epochNo: (0, Cardano_1.EpochNo)(epochNo),
            firstSlot: {
                date: slotTimeCalc((0, Cardano_1.Slot)(firstSlot)),
                slot: (0, Cardano_1.Slot)(firstSlot)
            },
            lastSlot: {
                date: slotTimeCalc((0, Cardano_1.Slot)(lastSlot)),
                slot: (0, Cardano_1.Slot)(lastSlot)
            }
        };
    };
};
exports.createSlotEpochInfoCalc = createSlotEpochInfoCalc;
