// @cardano-ogmios/client/dist/TxMonitor/awaitAcquire.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleAwaitAcquireResponse = exports.isAwaitAcquiredResult = exports.awaitAcquire = void 0;
const errors_1 = require("../errors");
const StateQuery_1 = require("../StateQuery");
const awaitAcquire = (context, args) => (0, StateQuery_1.Query)({
    methodName: 'AwaitAcquire',
    args: args
}, {
    handler: (response, resolve, reject) => {
        try {
            resolve((0, exports.handleAwaitAcquireResponse)(response));
        }
        catch (e) {
            reject(e);
        }
    }
}, context);
exports.awaitAcquire = awaitAcquire;
const isAwaitAcquiredResult = (result) => {
    if (typeof result !== 'object' || result === null) {
        return false;
    }
    return ('AwaitAcquired' in result && typeof result.AwaitAcquired === 'object');
};
exports.isAwaitAcquiredResult = isAwaitAcquiredResult;
const handleAwaitAcquireResponse = (response) => {
    const { result } = response;
    if ((0, exports.isAwaitAcquiredResult)(result)) {
        return result.AwaitAcquired.slot;
    }
    throw new errors_1.UnknownResultError(response);
};
exports.handleAwaitAcquireResponse = handleAwaitAcquireResponse;
