// @cardano-ogmios/client/dist/TxMonitor/nextTx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleNextTxResponse = exports.isNextTxResultAll = exports.isNextTxResultId = exports.nextTx = void 0;
const errors_1 = require("../errors");
const StateQuery_1 = require("../StateQuery");
function nextTx(context, args) {
    return (0, StateQuery_1.Query)({
        methodName: 'NextTx',
        args: args
    }, {
        handler: (response, resolve, reject) => {
            try {
                resolve(handleNextTxResponse(response, args));
            }
            catch (e) {
                reject(e);
            }
        }
    }, context);
}
exports.nextTx = nextTx;
const isNextTxResultId = (result) => (result === null) || (typeof result === 'string');
exports.isNextTxResultId = isNextTxResultId;
const isNextTxResultAll = (result) => (result === null) || (typeof result === 'object');
exports.isNextTxResultAll = isNextTxResultAll;
function handleNextTxResponse(response, args) {
    const { result } = response;
    if (args?.fields === 'all') {
        if ((0, exports.isNextTxResultAll)(result)) {
            return result;
        }
    }
    else {
        if ((0, exports.isNextTxResultId)(result)) {
            return result;
        }
    }
    throw new errors_1.UnknownResultError(response);
}
exports.handleNextTxResponse = handleNextTxResponse;
