// @cardano-ogmios/client/dist/TxMonitor/sizeAndCapacity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleSizeAndCapacityResponse = exports.isSizeAndCapacityResult = exports.sizeAndCapacity = void 0;
const errors_1 = require("../errors");
const StateQuery_1 = require("../StateQuery");
const sizeAndCapacity = (context, args) => (0, StateQuery_1.Query)({
    methodName: 'SizeAndCapacity',
    args: args
}, {
    handler: (response, resolve, reject) => {
        try {
            resolve((0, exports.handleSizeAndCapacityResponse)(response));
        }
        catch (e) {
            reject(e);
        }
    }
}, context);
exports.sizeAndCapacity = sizeAndCapacity;
const isSizeAndCapacityResult = (result) => {
    if (typeof result !== 'object' || result === null) {
        return false;
    }
    return ('capacity' in result && 'currentSize' in result && 'numberOfTxs' in result);
};
exports.isSizeAndCapacityResult = isSizeAndCapacityResult;
const handleSizeAndCapacityResponse = (response) => {
    const { result } = response;
    if ((0, exports.isSizeAndCapacityResult)(result)) {
        return result;
    }
    throw new errors_1.UnknownResultError(response);
};
exports.handleSizeAndCapacityResponse = handleSizeAndCapacityResponse;
