// @cardano-sdk/ogmios/dist/cjs/CardanoNode/OgmiosObservableCardanoNode/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ogmiosToCorePointOrOrigin = exports.ogmiosToCoreTipOrOrigin = exports.ogmiosToCoreTip = exports.ogmiosToCorePoint = exports.pointOrOriginToOgmios = void 0;
const core_1 = require("@cardano-sdk/core");
const pointOrOriginToOgmios = (point) => point === 'origin'
    ? 'origin'
    : {
        hash: point.hash,
        slot: point.slot
    };
exports.pointOrOriginToOgmios = pointOrOriginToOgmios;
const ogmiosToCorePoint = (point) => ({
    hash: core_1.Cardano.BlockId(point.hash),
    slot: core_1.Cardano.Slot(point.slot)
});
exports.ogmiosToCorePoint = ogmiosToCorePoint;
const ogmiosToCoreTip = (tip) => ({
    ...(0, exports.ogmiosToCorePoint)(tip),
    blockNo: core_1.Cardano.BlockNo(tip.blockNo)
});
exports.ogmiosToCoreTip = ogmiosToCoreTip;
const ogmiosToCoreTipOrOrigin = (tip) => (tip === 'origin' ? tip : (0, exports.ogmiosToCoreTip)(tip));
exports.ogmiosToCoreTipOrOrigin = ogmiosToCoreTipOrOrigin;
const ogmiosToCorePointOrOrigin = (point) => point === 'origin' ? point : (0, exports.ogmiosToCorePoint)(point);
exports.ogmiosToCorePointOrOrigin = ogmiosToCorePointOrOrigin;
