// @reef-defi/types/interfaces/definitions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "accounts", {
  enumerable: true,
  get: function () {
    return _definitions.default;
  }
});
Object.defineProperty(exports, "evm", {
  enumerable: true,
  get: function () {
    return _definitions2.default;
  }
});
Object.defineProperty(exports, "incentives", {
  enumerable: true,
  get: function () {
    return _definitions4.default;
  }
});
Object.defineProperty(exports, "poc", {
  enumerable: true,
  get: function () {
    return _definitions3.default;
  }
});
Object.defineProperty(exports, "primitives", {
  enumerable: true,
  get: function () {
    return _definitions5.default;
  }
});
Object.defineProperty(exports, "runtime", {
  enumerable: true,
  get: function () {
    return _definitions6.default;
  }
});
Object.defineProperty(exports, "support", {
  enumerable: true,
  get: function () {
    return _definitions7.default;
  }
});

var _definitions = _interopRequireDefault(require("./accounts/definitions"));

var _definitions2 = _interopRequireDefault(require("./evm/definitions"));

var _definitions3 = _interopRequireDefault(require("./poc/definitions"));

var _definitions4 = _interopRequireDefault(require("./incentives/definitions"));

var _definitions5 = _interopRequireDefault(require("./primitives/definitions"));

var _definitions6 = _interopRequireDefault(require("./runtime/definitions"));

var _definitions7 = _interopRequireDefault(require("./support/definitions"));
