// @cardano-sdk/ogmios/dist/cjs/CardanoNode/OgmiosCardanoNode.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _OgmiosCardanoNode_stateQueryClient, _OgmiosCardanoNode_logger, _OgmiosCardanoNode_connectionConfig;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OgmiosCardanoNode = void 0;
const core_1 = require("@cardano-sdk/core");
const client_1 = require("@cardano-ogmios/client");
const util_1 = require("@cardano-sdk/util");
const util_2 = require("../util");
const queries_1 = require("./queries");
class OgmiosCardanoNode extends util_1.RunnableModule {
    constructor(connectionConfig, logger) {
        super('OgmiosCardanoNode', logger);
        _OgmiosCardanoNode_stateQueryClient.set(this, void 0);
        _OgmiosCardanoNode_logger.set(this, void 0);
        _OgmiosCardanoNode_connectionConfig.set(this, void 0);
        __classPrivateFieldSet(this, _OgmiosCardanoNode_logger, (0, util_1.contextLogger)(logger, 'OgmiosCardanoNode'), "f");
        __classPrivateFieldSet(this, _OgmiosCardanoNode_connectionConfig, connectionConfig, "f");
    }
    async initializeImpl() {
        __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f").info('Initializing CardanoNode');
        __classPrivateFieldSet(this, _OgmiosCardanoNode_stateQueryClient, await (0, client_1.createStateQueryClient)(await (0, util_2.createInteractionContextWithLogger)(__classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f"), { connection: __classPrivateFieldGet(this, _OgmiosCardanoNode_connectionConfig, "f") })), "f");
        __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f").info('CardanoNode initialized');
    }
    async shutdownImpl() {
        __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f").info('Shutting down CardanoNode');
        await __classPrivateFieldGet(this, _OgmiosCardanoNode_stateQueryClient, "f").shutdown();
    }
    async eraSummaries() {
        if (this.state !== 'running') {
            throw new core_1.CardanoNodeErrors.NotInitializedError('eraSummaries', this.name);
        }
        return (0, queries_1.queryEraSummaries)(__classPrivateFieldGet(this, _OgmiosCardanoNode_stateQueryClient, "f"), __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f"));
    }
    async systemStart() {
        if (this.state !== 'running') {
            throw new core_1.CardanoNodeErrors.NotInitializedError('systemStart', this.name);
        }
        try {
            __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f").info('Getting system start');
            return await __classPrivateFieldGet(this, _OgmiosCardanoNode_stateQueryClient, "f").systemStart();
        }
        catch (error) {
            throw core_1.CardanoNodeUtil.asCardanoNodeError(error) || new core_1.CardanoNodeErrors.UnknownCardanoNodeError(error);
        }
    }
    async stakeDistribution() {
        if (this.state !== 'running') {
            throw new core_1.CardanoNodeErrors.NotInitializedError('stakeDistribution', this.name);
        }
        try {
            __classPrivateFieldGet(this, _OgmiosCardanoNode_logger, "f").info('Getting stake distribution');
            const map = new Map();
            for (const [key, value] of Object.entries(await __classPrivateFieldGet(this, _OgmiosCardanoNode_stateQueryClient, "f").stakeDistribution())) {
                const splitStake = value.stake.split('/');
                map.set(core_1.Cardano.PoolId(key), {
                    ...value,
                    stake: { pool: BigInt(splitStake[0]), supply: BigInt(splitStake[1]) }
                });
            }
            return map;
        }
        catch (error) {
            throw core_1.CardanoNodeUtil.asCardanoNodeError(error) || new core_1.CardanoNodeErrors.UnknownCardanoNodeError(error);
        }
    }
    async healthCheck() {
        try {
            const { networkSynchronization, lastKnownTip } = await (0, client_1.getServerHealth)({
                connection: (0, client_1.createConnectionObject)(__classPrivateFieldGet(this, _OgmiosCardanoNode_connectionConfig, "f"))
            });
            return {
                localNode: {
                    ledgerTip: lastKnownTip,
                    networkSync: core_1.Cardano.Percent(networkSynchronization)
                },
                ok: networkSynchronization > 0.99
            };
        }
        catch (error) {
            if (error.name === 'FetchError') {
                return { ok: false };
            }
            throw new core_1.ProviderError(core_1.ProviderFailure.Unknown, error);
        }
    }
    async startImpl() {
        return Promise.resolve();
    }
}
exports.OgmiosCardanoNode = OgmiosCardanoNode;
_OgmiosCardanoNode_stateQueryClient = new WeakMap(), _OgmiosCardanoNode_logger = new WeakMap(), _OgmiosCardanoNode_connectionConfig = new WeakMap();
