// @reef-defi/extension-inject/index.js
// Copyright 2019-2021 @polkadot/extension-inject authors & contributors
// SPDX-License-Identifier: Apache-2.0
export const REEF_EXTENSION_IDENT = 'reef';
export const REEF_INJECTED_EVENT = 'reef-injected'; // It is recommended to always use the function below to shield the extension and dapp from
// any future changes. The exposed interface will manage access between the 2 environments,
// be it via window (current), postMessage (under consideration) or any other mechanism

export function injectExtension(enable, {
  name,
  version
}) {
  // small helper with the typescript types, just cast window
  const windowInject = window;

  if (windowInject) {
    // don't clobber the existing object, we will add it (or create as needed)
    windowInject.injectedWeb3 = windowInject.injectedWeb3 || {}; // add our enable function

    windowInject.injectedWeb3[name] = {
      enable: origin => enable(origin),
      version
    };
  }
}
export function isInjected(name) {
  const windowInject = window;
  return !!(windowInject !== null && windowInject !== void 0 && windowInject.injectedWeb3) && !!(windowInject !== null && windowInject !== void 0 && windowInject.injectedWeb3[name]);
}
export function isInjectionStarted(name) {
  const windowInject = window;
  return !!windowInject._reefInjectionStart && !!windowInject._reefInjectionStart[name];
}
export function startInjection(name) {
  if (!window._reefInjectionStart) {
    window._reefInjectionStart = {};
  }

  window._reefInjectionStart[name] = true;
}
