// @cardano-sdk/core/dist/esm/util/nativeScript.js
import * as Cardano from '../Cardano';
import * as Crypto from '@cardano-sdk/crypto';
import * as toCml from '../CML/coreToCml';
import { SerializationError, SerializationFailure } from '../';
import { usingAutoFree } from '@cardano-sdk/util';
export const nativeScriptPolicyId = (script) => usingAutoFree((scope) => {
    const managedScript = toCml.nativeScript(scope, script);
    return Cardano.PolicyId(Buffer.from(scope.manage(managedScript.hash()).to_bytes()).toString('hex'));
});
export const jsonToNativeScript = (json) => {
    let coreScript;
    if (!json.type) {
        throw new SerializationError(SerializationFailure.InvalidScript, "Invalid Native Script. Missing 'type' field.");
    }
    switch (json.type) {
        case 'sig': {
            coreScript = {
                __type: Cardano.ScriptType.Native,
                keyHash: Crypto.Ed25519KeyHashHex(json.keyHash),
                kind: Cardano.NativeScriptKind.RequireSignature
            };
            break;
        }
        case 'all': {
            coreScript = {
                __type: Cardano.ScriptType.Native,
                kind: Cardano.NativeScriptKind.RequireAllOf,
                scripts: new Array()
            };
            for (let i = 0; i < json.scripts.length; ++i) {
                coreScript.scripts.push(jsonToNativeScript(json.scripts[i]));
            }
            break;
        }
        case 'any': {
            coreScript = {
                __type: Cardano.ScriptType.Native,
                kind: Cardano.NativeScriptKind.RequireAnyOf,
                scripts: new Array()
            };
            for (let i = 0; i < json.scripts.length; ++i) {
                coreScript.scripts.push(jsonToNativeScript(json.scripts[i]));
            }
            break;
        }
        case 'atLeast': {
            const required = Number.parseInt(json.required);
            coreScript = {
                __type: Cardano.ScriptType.Native,
                kind: Cardano.NativeScriptKind.RequireNOf,
                required,
                scripts: new Array()
            };
            for (let i = 0; i < json.scripts.length; ++i) {
                coreScript.scripts.push(jsonToNativeScript(json.scripts[i]));
            }
            break;
        }
        case 'before': {
            coreScript = {
                __type: Cardano.ScriptType.Native,
                kind: Cardano.NativeScriptKind.RequireTimeBefore,
                slot: Cardano.Slot(Number.parseInt(json.slot))
            };
            break;
        }
        case 'after': {
            coreScript = {
                __type: Cardano.ScriptType.Native,
                kind: Cardano.NativeScriptKind.RequireTimeAfter,
                slot: Cardano.Slot(Number.parseInt(json.slot))
            };
            break;
        }
        default: {
            throw new SerializationError(SerializationFailure.InvalidNativeScriptKind, `Native Script value '${json.type}' is not supported.`);
        }
    }
    return coreScript;
};
